[Clang] Apply exclude_from_explicit_instantiation to dllimport/dllexport

kikairoya opened the pull request:
Attaching `__declspec(dllexport/dllimport)` to explicit instantiation declaration made its whole member instantiated even if they were `__attribute__((__exclude_from_explicit_instantation__))`.
Such members should not be instantiated nor be exported to avoid symbol leakage or duplication.

Fixes #40363
Fixes #66909

kikairoya commented:
> That's great! Does this, together with #168170, address the overall problem in #135910, or are there many other issues remaining in order to fix that (other than reverting my old clang change)?

Neither of this nor #168170 is directly related to #135910 itself.
These are prerequisites for libc++'s ABI checking https://github.com/llvm/llvm-project/issues/135910#issuecomment-2901310436 , and another patch and checking functionality itself are still required to upstream.

For #135910 , I already have a "functional" local patch (with exporting nested classes), but I feel that I should try to find more edge cases.

zmodem commented:
> If I remember correctly, there are a number of preexisting references to the problem that `exclude_from_explicit_instantiation` doesn't do what it would need to do on Windows; I think such references may exist in libcxx (in the code tree maybe?) and/or as other issues here on github. It would be a nice extra if we could find some of those references, to get more stakeholders aware of this change here (and possibly know which references to update).

I think the main one is https://github.com/llvm/llvm-project/issues/40363 maybe. It would be good to check how this interacts with the problem described there.

kikairoya reviewed (commented):
I've added tests related to #40363 and #66909 .
Is it worth submitting these test-only patches first, to assert the current behavior?

zmodem commented on clang/test/CodeGenCXX/attr-exclude_from_explicit_instantiation.exclude_from_dllimport.cpp:
```diff
@@ -0,0 +1,169 @@
+// GNU: declare dllimport void @_ZN1DIiE14to_be_importedEv
+
+// MSC: define linkonce_odr dso_local void @"?also_to_be_imported@?$D@H@@QEAAXXZ"
+// GNU: define linkonce_odr dso_local void @_ZN1DIiE19also_to_be_importedEv
```
Hmm, this was a while ago so I don't have the whole context paged in, but I remember that it's tricky territory :-)

I think the problem was: how is it supposed to get exported?

Since you have an explicit instantiation decl for `D<int>`, you need a corresponding explicit instantiation definition on the exporting side. But if `D<>::also_to_be_imported` is `exclude_from_explicit_instantiation`, it will not get defined by that, so it doesn't get exported, and therefore cannot be imported.

rnk's comment is essentially saying that `dllimport` and `exclude_from_explicit_instantiation` have opposite purposes: `dllimport` says "this will be defined elsewhere", whereas `exclude_from_explicit_instantiation` says "this will *not* be explicitly defined elsewhere", and that the latter should probably win.

kikairoya commented on clang/test/CodeGenCXX/attr-exclude_from_explicit_instantiation.exclude_from_dllexport.cpp:
```diff
@@ -0,0 +1,171 @@
+// MSC: ModuleID = {{.*}}exclude_from_dllexport.cpp
+// MSC: source_filename = {{.*}}exclude_from_dllexport.cpp
+// GNU: ModuleID = {{.*}}exclude_from_dllexport.cpp
+// GNU: source_filename = {{.*}}exclude_from_dllexport.cpp
```
It's just to silence `--implicit-check=dllexport`.

zmodem replied:
Okay, but please add a comment near the top explaining how this test's use of `--implicit-check-not` works.

kikairoya commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -19108,8 +19118,23 @@ bool Sema::DefineUsedVTables() {
+            MarkFunctionReferenced(Loc, Method);
+            DefinedAnything = true;
+          }
+        }
```
I've rewritten here without changing behavior.

zmodem commented on clang/lib/Sema/SemaDeclCXX.cpp:
```diff
@@ -19109,7 +19119,17 @@ bool Sema::DefineUsedVTables() {
+            auto *Method = dyn_cast<CXXMethodDecl>(decl);
+            return Method && Method->isVirtual() &&
+                   Method->hasAttr<ExcludeFromExplicitInstantiationAttr>();
+          }))
```
As a style issue, I think it would be easier to read if we put this in a variable, like:

```
bool HasExcludeFromExplicitInstantiation = llvm::any_of(... hasAttr<ExcludeFromExlicitInstantiationAttr>() ...
```

Also, I'm not sure that it matters whether the method with the attribute is virtual or not. I think it would be enough to check that *any* method has the attribute -- because then the motivation about "suppress the vtable; it will live with the explicit instantiation definition" may not hold.

zmodem replied:
I wonder if we could break this out into a separate PR as well, since it seems somewhat separate from the other issue.

zmodem replied:
> I don't understand why non-virtual functions matter here.

A constructor method may trigger emitting the vtable. The larger point is that the reasoning on lines 19095-190198 about the "vtable will live with the explicit instantiation" doesn't necessarily hold if some methods are excluded from explicit instantiation.

zmodem replied:
> Why are non-virtual functions allowed to affect vtables?

When there's no key function, emitting a constructor will trigger emitting the vtable, because the constructor needs to set the vptr to point to it.

That's why the reasoning above about "vtable will live with the explicit instantiation" won't hold if the constructor is excluded.

kikairoya commented:
Gentle ping.

zmodem reviewed (commented):
Sorry, this got lost in the email pile over the holidays.

I'm still confused about the change to `Sema::DefineUsedVTables`. I'd rather we just check whether *any* method has `ExcludeFromExplicitInstantiationAttr`.

zmodem commented on clang/test/CodeGenCXX/attr-exclude_from_explicit_instantiation.exclude_from_dllexport.cpp:
```diff
@@ -0,0 +1,181 @@
+// MSC: define weak_odr dso_local dllexport{{.*}} ptr @"??4?$E@H@@QEAAAEAU0@$$QEAU0@@Z"
+// MSC: define weak_odr dso_local dllexport{{.*}} ptr @"??0?$E@H@@QEAA@XZ"
+// MSC: define weak_odr dso_local dllexport{{.*}} ptr @"??0?$E@H@@QEAA@AEBU0@@Z"
+// MSC: define weak_odr dso_local dllexport{{.*}} ptr @"??0?$E@H@@QEAA@$$QEAU0@@Z"
```
I don't think we need to check these implicitly generated functions.

zmodem commented on clang/test/CodeGenCXX/attr-exclude_from_explicit_instantiation.exclude_from_dllexport.cpp:
```diff
@@ -0,0 +1,181 @@
+void use() {
+  C<int> c;
+
+  // MSC: call void @"?to_be_exported_explicitly@?$C@H@@QEAAXXZ"
```
I don't think the expectations on the call instructions are needed. What's interesting is just if the function is dllexport or not.

zmodem reviewed (commented):
Thanks! I think the code looks good, but I think we can simplify the tests. I've focused on the first file, but the comments apply to both.

kikairoya reviewed (commented):
I'm sorry for my delay.

Some of the feedback relates to use of `--implicit-check-not`, but I think`--implicit-check-not` is necessary to catch unintended imports/exports, which is important since this PR fixes the unexpected imports/exports.
Something like `--implicit-check-not-regex="(define|declare).*to_be_"` could reduce redundant checks. What do you think about adding such a feature to `lit`?

zmodem reviewed (commented):
Thanks for bearing with me.

I still wish these tests could be less complex. The functionality change isn't that complicated, but the tests seem to have hundreds of checks that will be hard to maintain.

Can't we just have a couple of classes with explicit checks for the stuff that's changing with this patch?

kikairoya commented on clang/test/CodeGenCXX/attr-exclude_from_explicit_instantiation.exclude_from_dllexport.cpp:
```diff
@@ -0,0 +1,179 @@
+
+  // This will be instantiated implicitly as an exported function because it is
+  // marked as dllexport explicitly.
+  EXCLUDE_ATTR __declspec(dllexport) void to_be_memberwise_exported();
```
A `dllexport`ed explicit instantiation triggers instantiation of the method but a non-`dllexport`ed explicit instantiation doesn't instantiate it.

I think a memberwise-`dllimport`ed method should be imported regardless whether it is excluded or how it was instantiated, then memberwise-`dllexport`ed method should be exported even if the explicit instantiation definition were not exported. Does that makes sense?

kikairoya replied:
Or, would it be better to just reject the combination of `exclude_from_explicit_instantiation` and `dllexport`/`dllimport`?
That might make the semantics simpler, like "excluded members would never be exported/imported", and align with the intent of #40363 .

zmodem replied:
Hmm, the rabbit hole goes deeper :-)

In my experience, dllimport/export attributes on members is much less common than putting it on classes, but we should of course strive to get the behavior right.

I'm not sure we want to /reject/ the combination of exclude_from_explicit_instantiation and dllimport/export on a member, but how about we drop (ignore) the dllimport/export attribute in that case (with a warning)?

That would be consistent with how we handle class-level dllimport/export: it gets ignored for methods with exclude_from_explicit_instantiation. And it fits with the intent of #40363.

zmodem reviewed (commented):
> That change consists of two steps: make the exclude attribute ignored on a non-template context, and make the exclude attribute takes precedence over the DLL attributes.
> If this approach makes sense, I'll submit them as separated PRs.

Thanks! That looks great, and doing it as a separate PR sounds good to me.

kikairoya committed:
pretest

kikairoya committed:
[Clang] Apply exclude_from_explicit_instantiation to dllimport/dllexport

Attaching `__declspec(dllexport/dllimport)` to explicit instantiation
declaration made its whole member instantiated even if they were
`__attribute__((__exclude_from_explicit_instantation__))`.
Such members should not be instantiated nor be exported to avoid
symbol leakage or duplication.

kikairoya committed:
instantiate vtable

kikairoya commented on clang/lib/Sema/SemaDeclCXX.cpp:
There was a check which has been dropped now:

```c++
      if ((TSK == TSK_ExplicitInstantiationDeclaration ||
           TSK == TSK_ExplicitInstantiationDefinition) &&
           Member->hasAttr<ExcludeFromExplicitInstantiationAttr>())
        continue;
```

since any member shouldn't have both the exclude attribute and the dllimport/dllexport attributes.

kikairoya committed:
not to use temporaries in %t.dir

kikairoya committed:
exclude-attr affects class-level dll-attr also

kikairoya committed:
omit redundant noinline checks

zmodem reviewed (approved):
lgtm
